### DISCUSSION TITLE
InOneWeekend: issue with dielectrics?

### DISCUSSION THREAD

**Discussion by mariano-gn** [RayTracing/raytracing.github.io#769] [Debugging]
Hello!

I tried porting RTIOW to GLSL, fun ride! But I found what it seems to be an issue with dielectrics that I don't understand. I also downloaded the reference implementation from the repo (5c312de55e33fa97b0885b00ec86c908f7adba0b) and it appears to suffer from the same problem?

Here's the patches I used to repro a set of 'similar' scenes: 
[dielectrics-issue-diff.txt](https://github.com/RayTracing/raytracing.github.io/files/5393721/dielectrics-issue-diff.txt)
[dielectrics-issue-diff-2.txt](https://github.com/RayTracing/raytracing.github.io/files/5393722/dielectrics-issue-diff-2.txt)
[dielectrics-issue-diff-3.txt](https://github.com/RayTracing/raytracing.github.io/files/5393723/dielectrics-issue-diff-3.txt)

Here are my results (issue highlighted):
![dielectrics-issue-glsl](https://user-images.githubusercontent.com/8467452/96297057-974e7080-0ff0-11eb-9acb-ec4db5918918.png)
![dielectrics-issue-glsl-2](https://user-images.githubusercontent.com/8467452/96297074-9ae1f780-0ff0-11eb-87e9-2aeafef4b6ef.png)
![dielectrics-issue-glsl-3](https://user-images.githubusercontent.com/8467452/96297076-9b7a8e00-0ff0-11eb-8b45-77be5811d7a5.png)

And here are the reference results (issue highlighted):
![dielectrics-issue-reference](https://user-images.githubusercontent.com/8467452/96297099-a6352300-0ff0-11eb-8a6e-a1591d335935.png)
![dielectrics-issue-reference-2](https://user-images.githubusercontent.com/8467452/96297107-a9301380-0ff0-11eb-8293-74f40b42676d.png)
![dielectrics-issue-reference-3](https://user-images.githubusercontent.com/8467452/96297110-a9c8aa00-0ff0-11eb-913a-23a86588573a.png)

I'll try to understand what's really going on, but I wanted to rise the issue now. Since I don't have glass/metal balls to play with I'll have to do it on the actual code :)

Cheers,

**Comment by hollasch** [comment]
Would be useful to render a glass sphere behind a glass sphere (the point of view of the golden sphere on the right. I'd do three renders: one slightly to the left, one directly in line, and one slightly to the right.

**Comment by trevordblack** [comment]
The most probable thing that I'm seeing is that the rays bouncing of the metal sphere are only seeing 1 of the two spheres.

The bounces closer to camera see the middle glass sphere, and the bounces further from camera see the left glass sphere. Maybe check to make sure that you are traversing the scene correctly? You might be having a variable that has a too-small scope.

Otherwise:
How are you doing bounces in glsl? is it recursion (unlikely), or looping? It looks like you're either
1. running out of bounces
2. You're at a floating point rounding boundary

**Comment by mariano-gn** [comment]
Hi! Thank you for your answers, I'm not doing much with the project this week, but I'll try to come back to it on the weekend. Trying to answer your comments already:

> Would be useful to render a glass sphere behind a glass sphere (the point of view of the golden sphere on the right. I'd do three renders: one slightly to the left, one directly in line, and one slightly to the right.

Did that, (don't have the render at hand but I'll try again) _but_ I rendered it dead on the position of the metal sphere on the right (and tbh, nothing was glaringly wrong, but I didn't have a couple glass spheres to compare to, will also try the same with the reference implementation). Will try to either render it statically from the positions you mention and/or, if possible with a slight movement, since when using GLSL is entirely possible to do it at a not-so-terrible speed. 👍 

> The most probable thing that I'm seeing is that the rays bouncing of the metal sphere are only seeing 1 of the two spheres.

It does look like that yes, but the strangest thing is:
- When rendered only the left-most sphere + metal, it's relatively small compared to any other render.
- When rendered only the middle sphere + metal, the reflection looks relatively reasonable.
- When rendering the 3, you can see 2 different variants of the middle sphere, the one that looks relatively reasonable and that other ghost reflection with weird looks.

> The bounces closer to camera see the middle glass sphere, and the bounces further from camera see the left glass sphere. 

Not entirely sure it's that, the bounces further from the camera are more similar to the rendering-only-the-middle-sphere, and the ones closer to the camera look plain wrong.

> Maybe check to make sure that you are traversing the scene correctly? You might be having a variable that has a too-small scope.

Good points, in GLSL I'm just having a list of objects, no BVH of any kind, yet small scope variables could be a problem yes.

> Otherwise:
> How are you doing bounces in glsl? is it recursion (unlikely), or looping? It looks like you're either

It's looping with tail call-like replacement, took inspiration from here:  https://www.cs.uaf.edu/2012/spring/cs481/section/0/lecture/02_07_recursion_reflection.html

> 1. running out of bounces

Can try increasing bounces, great idea!
> 2. You're at a floating point rounding boundary

Interesting and yeah... Possible too! 👍 

Now, I just want to point out that recursion issues were one of my first suspects, so I tried to reproduce the scene using the reference renderer: I cloned _this_ repo and tried to reproduce the similar scene as best as I could (see my patches + images corresponding to _this_ implementation).

What I meant with `reference` is just that, I cloned this repo, modified the world generation function and tried to render the 'same' scene but using C++ instead of GLSL.

Thanks again for the suggestions!
